frame/bridge/relayer-game/src/lib.rs

```diff
@@ -0,0 +1,688 @@
+use sp_runtime::{
+	traits::{SaturatedConversion, Saturating, Zero},
+	RuntimeDebug,
+};
+#[cfg(not(feature = "std"))]
+use sp_std::borrow::ToOwned;
+use sp_std::{collections::btree_map::BTreeMap, prelude::*};
+// --- darwinia ---
+use darwinia_support::{balance::lock::*, relay::*};
+use types::*;
+
+const RELAYER_GAME_ID: LockIdentifier = *b"da/rgame";
+
+pub trait Trait<I: Instance = DefaultInstance>: frame_system::Trait {
+	type Event: From<Event<Self, I>> + Into<<Self as frame_system::Trait>::Event>;
+
+	/// The currency use for bond
+	type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+
+	/// Handler for the unbalanced *RING* reduction when slashing a relayer.
+	type RingSlash: OnUnbalanced<RingNegativeImbalance<Self, I>>;
+
+	/// A regulator to adjust relay args for a specific chain
+	type RelayerGameAdjustor: AdjustableRelayerGame<
+		Balance = RingBalance<Self, I>,
+		Moment = Self::BlockNumber,
+		TcBlockNumber = TcBlockNumber<Self, I>,
+	>;
+
+	/// The target chain's relay module's API
+	// TODO: MMR type
+	type TargetChain: Relayable<TcHeaderMMR = ()>;
+}
+
+decl_event! {
+	pub enum Event<T, I: Instance = DefaultInstance>
+	where
+		AccountId = AccountId<T>,
+		BlockNumber = BlockNumber<T>,
+	{
+		/// TODO
+		TODO(AccountId, BlockNumber),
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait<I>, I: Instance> {
+		/// Challenge - NOT HAPPENED
+		ChallengeNH,
+
+		/// Target Header - ALREADY CONFIRMED
+		TargetHeaderAC,
+
+		/// Proposal - ALREADY EXISTED
+		ProposalAE,
+		/// Target Header - ALREADY EXISTED
+		TargetHeaderAE,
+
+		/// Round - MISMATCHED
+		RoundMis,
+
+		/// Can not bond with value less than usable balance.
+		InsufficientValue,
+	}
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait<I>, I: Instance = DefaultInstance> as DarwiniaRelayerGame {
+		/// Each target chain's header relay can open a game
+		pub Proposals
+			get(fn proposals_of_game)
+			: map hasher(blake2_128_concat) GameId<TcBlockNumber<T, I>>
+			=> Vec<Proposal<
+				AccountId<T>,
+				BondedTcHeader<
+					RingBalance<T, I>,
+					TcHeaderBrief<TcBlockNumber<T, I>, TcHeaderHash<T, I>, ()>
+				>,
+				TcHeaderHash<T, I>
+			>>;
+
+		/// The allow samples for each game
+		pub Samples
+			get(fn samples_of_game)
+			: map hasher(blake2_128_concat) TcBlockNumber<T, I>
+			=> Vec<TcBlockNumber<T, I>>;
+
+		/// The closed rounds which had passed the challenge time at this moment
+		pub ClosedRounds
+			get(fn closed_rounds_at)
+			: map hasher(blake2_128_concat) BlockNumber<T>
+			=> Vec<(GameId<TcBlockNumber<T, I>>, Round)>;
+
+		pub Bonds
+			get(fn bond_of_relayer)
+			: map hasher(blake2_128_concat) AccountId<T>
+			=> RingBalance<T, I>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait<I>, I: Instance = DefaultInstance> for enum Call
+	where
+		origin: T::Origin
+	{
+		type Error = Error<T, I>;
+
+		fn deposit_event() = default;
+
+		// TODO: too many db operations and calc need to move to `offchain_worker`
```

Comment:
Need some docs for the parts:

1. on_chain_arbitrate?
2. update_samples?
3. settle_without_challenge?

Could they be merged to child functions?